examples/balzano/balzano.py

```diff
@@ -0,0 +1,113 @@
+#
+# In addition to bathymetry, elevation and velocity fields,
+# user-specified outputs are used for the moving bathymetry
+# (h_tilde) and total water depth imposed on original
+# bathymetry (eta_tilde), the latter being useful for
+# comparisons with other WD models.
+#
+# [1] O. Gourgue, R. Comblen, J. Lambrechts, T. Karna, V.
+#     Legat, and E. Deleersnijder. A flux-limiting wetting-
+#     drying method for finite-element shallow-water models,
+#     with application to the scheldt estuary. Advances in
+#     Water Resources, 32:1726 - 1739, 2009.
+#     doi: 10.1016/j.advwatres.2009.09.005.
+#
+# Simon Warder 2017-03-21
+
+from thetis import *
+
+outputdir = 'outputs'
+mesh2d = RectangleMesh(12, 6, 13800, 7200)
+print_output('Exporting to '+outputdir)
+
+# time step in seconds
+dt = 600.
+# total duration in seconds: 4 periods
+t_end = 2*24*3600.
+# export interval in seconds
+t_export = 600.
+
+# bathymetry: uniform slope with gradient 1/2760
+P1_2d = FunctionSpace(mesh2d, 'CG', 1)
+bathymetry = Function(P1_2d, name='Bathymetry')
+x = SpatialCoordinate(mesh2d)
+bathymetry.interpolate(x[0] / 2760.0)
+
+# bottom friction suppresses reflection from wet-dry front
+mu_manning = Constant(0.02)
+# wetting-drying options
+wetting_and_drying = True
+wd_alpha = Constant(0.4)
+
+# --- create solver ---
+solverObj = solver2d.FlowSolver2d(mesh2d, bathymetry)
+options = solverObj.options
+options.t_export = t_export
+options.t_end = t_end
+options.outputdir = outputdir
+options.check_vol_conservation_2d = True
+options.fields_to_export = ['uv_2d', 'elev_2d']
+options.timestepper_type = 'cranknicolson'
+options.shallow_water_theta = 0.5
+options.wetting_and_drying = wetting_and_drying
+options.wd_alpha = wd_alpha
+options.mu_manning = mu_manning
+options.dt = dt
+options.solver_parameters_sw = {
+    'snes_type': 'newtonls',
+    'snes_monitor': True,
+    'ksp_type': 'gmres',
+    'pc_type': 'fieldsplit',
+}
+
+# boundary conditions
+h_amp = -2.0      # ocean boundary forcing amplitude
+h_T = 12*3600.    # ocean boundary forcing period
+ocean_elev_func = lambda t: h_amp * sin(2 * pi * t / h_T)
+solverObj.create_function_spaces()
+H_2d = solverObj.function_spaces.H_2d
+ocean_elev = Function(H_2d, name="ocean boundary elevation").assign(ocean_elev_func(0.))
+ocean_funcs = {'elev': ocean_elev}
+solverObj.bnd_functions['shallow_water'] = {
+    2: ocean_funcs
+}
+
+# User-defined output: moving bathymetry and eta_tilde
+wd_bathfile = File(os.path.join(outputdir, 'moving_bath.pvd'))
+moving_bath = Function(P1_2d, name="moving_bath")
+eta_tildefile = File(os.path.join(outputdir, 'eta_tilde.pvd'))
+eta_tilde = Function(P1_2d, name="eta_tilde")
+
+
+# user-specified export function
+def export_func():
+    wd_bath_displacement = solverObj.eq_sw.bathymetry_displacement_mass_term.wd_bathymetry_displacement
+    eta = solverObj.fields.elev_2d
+    moving_bath.project(bathymetry + wd_bath_displacement(eta))
+    wd_bathfile.write(moving_bath)
```

Comment:
Hmm, the right way to export these new fields this is to use the exporting functionality: add the field metadata to `field_defs.py` and then users can export the fields just by adding `'moving_bath_2d'` (or such) to `options.fields_to_export` list. It will provide all supported output formats.

In this case, however, we need to call the `project` somewhere. So I think we need to add a mechanism to `Exporter` class to support evaluation callbacks. We'd probably need to rewrite parts of the exporter functionality (I'm not a huge fan of `fields_defs` for example). Thus I suggest that we leave this example as it is and address it in another PR.